src/shared/clients/iotClient.ts

```diff
@@ -0,0 +1,702 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import * as _ from 'lodash'
+import * as mime from 'mime-types'
+import * as path from 'path'
+import { AWSError, Iot } from 'aws-sdk'
+import { inspect } from 'util'
+import { ext } from '../extensionGlobals'
+import { getLogger } from '../logger'
+import { DefaultFileStreams, FileStreams, pipe, promisifyReadStream } from '../utilities/streamUtilities'
+import { InterfaceNoSymbol } from '../utilities/tsUtils'
+
+export const DEFAULT_MAX_THINGS = 250 // 250 is the maximum allowed by the API
+export const DEFAULT_DELIMITER = '/'
+
+export type IotThing = InterfaceNoSymbol<DefaultIotThing>
+export type IotCertificate = InterfaceNoSymbol<DefaultIotCertificate>
+export type IotPolicy = InterfaceNoSymbol<DefaultIotPolicy>
+export type IotClient = InterfaceNoSymbol<DefaultIotClient>
+
+//ARN Pattern for certificates. FIXME import @aws-sdk/util-arn-parser instead.
+const CERT_ARN_PATTERN = /arn:aws:iot:\S+?:\d+:cert\/(\w+)/
+
+interface IotObject {
+    readonly key: string
+    readonly versionId?: string
+}
+
+export interface ListThingsRequest {
+    readonly nextToken?: Iot.NextToken
+    readonly maxResults?: Iot.MaxResults
+    readonly principal?: Iot.Principal
+}
+
+export interface ListThingsResponse {
+    readonly things: IotThing[]
+    readonly nextToken: string | undefined
+}
+
+export interface UpdateThingRequest {
+    readonly thingName: Iot.ThingName
+}
+
+export interface CreateThingResponse {
+    readonly thing: IotThing
+}
+
+export interface ListCertificatesRequest {
+    readonly pageSize?: Iot.PageSize
+    readonly marker?: Iot.Marker
+    readonly ascendingOrder?: Iot.AscendingOrder
+}
+
+export interface ListCertificatesResponse {
+    readonly certificates: IotCertificate[]
+    readonly nextMarker: string | undefined
+}
+
+export interface ListThingCertificatesRequest {
+    readonly thingName: Iot.ThingName
+    readonly nextToken?: Iot.NextToken
+    readonly maxResults?: Iot.MaxResults
+}
+
+export interface ListThingCertificatesResponse {
+    readonly certificates: IotCertificate[]
+    readonly nextToken: string | undefined
+}
+
+export interface UpdateCertificateRequest {
+    readonly certificateId: Iot.CertificateId
+    readonly newStatus: Iot.CertificateStatus
+}
```

Comment:
No need for these wrapper request and response types - we can just interact directly with the underlying IOT SDK stuff. Generally these wrapper clients do not add very much value for us (something I want to clean up at some point) so let's keep it as lightweight as possible.

Reply:
I removed most of these, except for the ones I added that aren't part of the IoT SDK (i.e. the ones that take extra parameters in the form of filepaths on the local filesystem). How should I handle those ones?